### ISSUE TITLE
Weed out non-existent dirs?

### ISSUE THREAD

**Issue by xendk** [stevenremot/ede-php-autoload#21]
Now that #19 merges paths, should there be a `:late` visitor that removes non-existent paths? They don't do much harm, but if composer/installers has relocated a lot of projects, it might be a bit ineffective to repeatedly check if classes exists in non-existing dirs.

**Comment by stevenremot** [comment]
For this one I would rather keep it a bit like this, because I'm afraid of practical consequences, like cloning a project, opening it in Emacs, and doing `composer install` only after that. Because you started emacs before the vendor directory is created, that may cause problems... Or not...

Anyway, I would like to wait a bit for people to use it before removing non-existing directories. What do you think ?

**Comment by xendk** [comment]
Well, yeah, the Emacs before composer case would be a problem. But there's some other possible problems related to that, that might provide a possible solution. Doing `git pull && composer install` in a project already loaded in Emacs wouldn't discover any new packages, I believe. And if there's no `composer.lock` when Emacs starts, I think it can't find the autoloaded namespaces anyway?

It would be handy if ede-php-autoload could detect that composer has run and refresh the data. Unfortunately, composer doesn't touch `composer.lock` if there's no changes to that. It does, however, touch `vendor/autoload.php`, whenever it is run, regardless of whether it actually installed anything. So the mtime of that file could serve as a signal for reloading the autoloads.

I think that would deal with all three mentioned issues. Regardless, an `ede-php-autoload-composer-refresh` command could be handy for triggering a manual reload. Gives people something to try, in case of problems.

**Comment by stevenremot** [comment]
Mmmmh, you're right. The best would be to come up with a refresh command before handling this issue. I will create a new issue for this. However I will pause a bit the development of this package for a few days now that it's functionnal, I want to work on other projects too  :-)

**Comment by xendk** [comment]
Yeah, I should get ede-php-autoload-drupal release ready too. :-)

**Comment by xendk** [comment]
I just got burned by this. I'm investigating creating an ivy completion for opening a file by class name (you know, scratching your own itch), but `ede-php-autoload-complete-type-name` fails when it hits a project that ede-autoload-composer-installers has relocated because it tries to read the nonexistent dir the default visitor added.

I'll try to do a visitor that does that, but it could live in ede-autoload-composer-installers, as that's about the only one that needs it at the moment. What do you think?

**Comment by stevenremot** [comment]
I was thinking about integrating this cleaning phase directly in ede-php-autoload. It would also filter autoloads you created by hand on filesdirectories that doesn't exist then.

And now that I implemented the "reload autoloads" features, I'm more willing to do this 🙂

**Comment by xendk** [comment]
Sounds like a better plan to me.

**Comment by stevenremot** [comment]
Done in the PR https://github.com/stevenremot/ede-php-autoload/pull/25 . I merge it as soon as it's OK for you

**Comment by xendk** [comment]
:+1: 

Now `ede-php-autoload-complete-type-name` can traverse `Drupal\Core`. :-)

**Comment by stevenremot** [comment]
Merged!
